3/assets/3.6-finished-code.rs

```diff
@@ -0,0 +1,315 @@
+use parity_codec::Encode;
+use support::{decl_storage, decl_module, StorageValue, StorageMap,
+    dispatch::Result, ensure, decl_event};
+use system::ensure_signed;
+use runtime_primitives::traits::{As, Hash, Zero};
+use rstd::cmp;
+
+#[derive(Encode, Decode, Default, Clone, PartialEq)]
+pub struct Kitty<Hash, Balance> {
+    id: Hash,
+    dna: Hash,
+    price: Balance,
+    gen: u64,
+}
+
+#[derive(Encode, Decode, Default, Clone, PartialEq, Debug)]
+pub struct Auction<Hash, Balance, BlockNumber, AccountId> {
+    kitty_id: Hash,
+    kitty_owner: AccountId,
+    expiry: BlockNumber,
+    min_bid: Balance,
+    high_bid: Balance,
+    high_bidder: AccountId,
+}
+
+pub trait Trait: balances::Trait {
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+}
+
+decl_event!(
+    pub enum Event<T>
+    where
+        <T as system::Trait>::AccountId,
+        <T as system::Trait>::Hash,
+        <T as balances::Trait>::Balance,
+        <T as system::Trait>::BlockNumber
+    {
+        Created(AccountId, Hash),
+        PriceSet(AccountId, Hash, Balance),
+        Transferred(AccountId, AccountId, Hash),
+        Bought(AccountId, AccountId, Hash, Balance),
+        Auction(Hash, Balance, BlockNumber), // kitty_id, min_bid, expiry
+        Bid(Hash, Balance, AccountId), // kitty_id, high_bid, bidder
```

Comment:
There should be an `AuctionFinalized` event which is emitted from the `on_finalise` logic